File: weekday/src/main/java/com/team3/weekday/action/ProjectAction.java
package com.team3.weekday.action;

import com.team3.weekday.IDataBus;
import com.team3.weekday.constant.CommonConstant;
import com.team3.weekday.net.ChannelSession;
import com.team3.weekday.net.annotation.Action;
import com.team3.weekday.net.annotation.Command;
import com.team3.weekday.net.annotation.Param;
import com.team3.weekday.utils.ResponseUtil;
import com.team3.weekday.valueobject.Response;
import com.team3.weekday.valueobject.TaskStatus;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Date;

@Action
public class ProjectAction {

    @Autowired
    private IDataBus dataBus;

    @Command
    public Response createProject(@Param Long teamId,
                                  @Param String name,
                                  ChannelSession session) throws Exception {
        if (teamId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_TEAM_EMPTY);
        }

        return dataBus.projectService().createProject(session.getUserId(), teamId, name);
    }

    @Command
    public Response updateProject(@Param Long projectId,
                                  @Param String name,
                                  @Param String background,
                                  @Param String description,
                                  ChannelSession session) throws Exception {
        if (projectId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_PROJECT_EMPTY);
        }
        return dataBus.projectService().updateProject(session.getUserId(), projectId, name, background, description);
    }

    @Command
    public Response deleteProject(@Param Long projectId,
                                  ChannelSession session) throws Exception {
        if (projectId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_PROJECT_EMPTY);
        }
        return dataBus.projectService().deleteProject(session.getUserId(), projectId);
    }

    @Command
    public Response getProjectInfo(@Param Long projectId,
                                   ChannelSession session) throws Exception {
        if (projectId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_PROJECT_EMPTY);
        }
        return dataBus.projectService().getProjectInfo(projectId);
    }

    @Command
    public Response createTask(@Param Long projectId,
                               @Param String name,
                               @Param String content,
                               @Param Long ownerId,
                               @Param Long requesterId,
                               @Param Date start,
                               @Param Date end,
                               @Param Integer complexity,
                               ChannelSession session) throws Exception {
        if (projectId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_PROJECT_EMPTY);
        }
        return dataBus.projectService().createTask(projectId, name, content, ownerId, requesterId, start, end,
                null, complexity);
    }

    @Command
    public Response createSubTask(@Param String name,
                                  @Param Long ownerId,
                                  @Param Date start,
                                  @Param Date end,
                                  @Param Long masterId,
                                  ChannelSession session) throws Exception {
        if (masterId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_TASK_EMPTY);
        }

        return dataBus.projectService().createSubTask(name, ownerId, start, end, masterId);
    }

    @Command
    public Response updateTask(@Param Long taskId,
                               @Param String name,
                               @Param String content,
                               @Param Long ownerId,
                               @Param Long requesterId,
                               @Param Date start,
                               @Param Date end,
                               @Param Integer type,
                               @Param Long masterId,
                               @Param Integer complexity,
                               ChannelSession session) throws Exception {
        if (taskId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_TASK_EMPTY);
        }

        return dataBus.projectService().updateTask(taskId, name, content, ownerId, requesterId, start, end,
                type, masterId, complexity);
    }

    @Command
    public Response updateTaskStatus(@Param Long taskId,
                                     @Param Integer status,
                                     ChannelSession session) throws Exception {
        if (taskId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_TASK_EMPTY);
        }
        TaskStatus taskStatus = TaskStatus.getTaskStatus(status);
        if (taskStatus == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.PARAMETER_IS_WRONG, "taskStatus");
        }

        return dataBus.projectService().updateTaskStatus(taskId, taskStatus);
    }

    @Command
    public Response deleteTask(@Param Long taskId,
                               ChannelSession session) throws Exception {
        if (taskId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_TASK_EMPTY);
        }
        return dataBus.projectService().deleteTask(taskId);
    }

    @Command
    public Response getComments(@Param Long taskId,
                                ChannelSession session) throws Exception {
        if (taskId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_TASK_EMPTY);
        }
        return dataBus.projectService().getComments(session, taskId);
    }

    @Command
    public Response addComment(@Param Long taskId,
                               @Param String content,
                               ChannelSession session) throws Exception {
        if (taskId == null) {
            return ResponseUtil.sendErrorMessage(CommonConstant.ERROR_TASK_EMPTY);
        }
        return dataBus.projectService().addComment(session, taskId, content);
    }
}

File: weekday/src/main/resources/static/js/action/projectAction.js
import { socket_send } from "../net/socketConnect.js";
import { checkMemberId } from "../common/commonService.js";

function createProject(id, name) {
    let message = {"cmd": "createProject", "teamId": id, "name": name.toString()}
    socket_send(message);
}

function updateProject(id) {
    let project = $('#update-project')
    let message = {
        "cmd": "updateProject", "projectId": id, "name": project.find('#name').val(),
        "description": project.find('#description').val()
    }
    socket_send(message)
}

function deleteProject(id) {
    let message = {"cmd": "deleteProject", "projectId": id}
    socket_send(message)
}

function createTask(projectId) {
    let owner = $('#task-owner').val()
    let requester = $('#task-requester').val()
    let teamId = $('#member-list').attr('teamid')
    owner = checkMemberId(owner, teamId)
    requester = checkMemberId(requester, teamId)
    let message = {
        "cmd": "createTask", "projectId": projectId.toString(), "name": $("#task-name").val(),
        "content": $("#task-content").val(), "ownerId": owner, "requesterId": requester,
        "start": $("#task-start").val(), "end": $("#task-end").val(), "complexity": $("#task-complex").val()
    }
    socket_send(message)
}

function updateTask(id, teamId) {
    let table = $('#update-table')
    let owner = table.find('#owner').val()
    owner = checkMemberId(owner, teamId)
    let message = {
        "cmd": "updateTask", "taskId": id, "name": table.find('#name').val(), "content": table.find('#content').val(),
        "ownerId": owner, "start": table.find('#start').val(), "end": table.find('#end').val(),
        "complexity": table.find('#complexity').val()
    }
    socket_send(message)
}

function addComment(comment) {
    let id = $('#show-comment').attr('taskid')
    let message = {"cmd": "addComment", "taskId": id, "content": comment.toString()}
    socket_send(message)
}

function getComment(id) {
    let message = {"cmd": "getComments", "taskId": id.toString()}
    socket_send(message)
}

function updateTaskStatus(id, content) {
    let status = Number(content) + 1
    let message = {"cmd": "updateTaskStatus", "taskId": id.toString(), "status": status.toString()}
    socket_send(message)
}

function createSubTask(id) {
    let task = $('#sub-table')
    let owner = task.find('#owner').val()
    let teamId = $('#member-list').attr('teamid')
    owner = checkMemberId(owner, teamId)
    let message = {
        "cmd": "createSubTask", "masterId": id.toString(), "name": task.find("#name").val(),
        "ownerId": owner, "start": task.find('#start').val(), "end": task.find('#end').val()
    }
    socket_send(message)
}

function deleteTask(id) {
    let message = {"cmd": "deleteTask", "taskId": id}
    socket_send(message)
}

export { createProject, createTask, createSubTask }
export { updateProject, updateTask, addComment, getComment, updateTaskStatus }
export { deleteProject, deleteTask }